### PR TITLE
feat(notifications): add notifications domain project

### PR DESCRIPTION
**Pull request by osk**
# Add notifications domain

* Connection to notification client
* With mocked response to get the GraphQL API out there for clients (service portal + app) to develop against.

<details><summary>Example list query</summary>

```
query Notifications {
    notifications(input: { first: 10 }) {
        data {
            notificationId
            metadata {
                sent
                read
                status
            }
            sender {
                name
                logo
            }
            recipient {
                nationalId
            }
            message {
                title
                body
                link {
                    uri
                }
            }
        }
        totalCount
        messageCounts {
            totalCount
            unreadCount
        }
        pageInfo {
            endCursor
            hasNextPage
            hasPreviousPage
            startCursor
        }
    }
}
```

</details> 

<details><summary>Example detail query</summary>

```
query Notification {
    notification(notificationId: "<id>") {
        data {
            notificationId
            metadata {
                sent
                read
                status
            }
            sender {
                name
                logo
            }
            recipient {
                nationalId
            }
            message {
                title
                body
                link {
                    uri
                }
            }
        }
    }
}

```

</details> 

<details><summary>Example mark as read mutation</summary>

```
mutation MarkNotificationAsRead {
    markNotificationAsRead(notificationId: "<id>") {
        data {
            notificationId
            metadata {
                sent
                read
                status
            }
            sender {
                name
                logo
            }
            recipient {
                nationalId
            }
            message {
                title
                body
                link {
                    uri
                }
            }
        }
    }
}

```

</details> 

[See Notion documention](https://www.notion.so/Hnipp-cf817ec3b38e4b55a013786f364e90b9).

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/api/domains/notifications/src/lib/notifications.model.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  sender!: NotificationSender
+
+  @Field(() => NotificationRecipient)
+  recipient!: NotificationRecipient
+
+  @Field(() => NotificationMessage)
+  message!: NotificationMessage
+}
+
+@ObjectType('MessageCounts')
```
Comment (thorkellmani) [inline libs/api/domains/notifications/src/lib/notifications.model.ts:82]: Must prefix all GraphQL object types with the domain name unless the object type contains either singular or plural version of the domain name, see [Code docs](https://docs.devland.is/technical-overview/api-design-guide/graphql-naming-conventions)
```suggestion
@ObjectType('NotificationsMessageCounts')
```

File: libs/api/domains/notifications/src/lib/notifications.resolver.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        notificationId,
+        locale,
+        category: LOG_CATEGORY,
+      })
+      throw e
+    }
+
+    if (!result) {
+      throw new NotFoundException('notification not found')
+    }
```
Comment (thorkellmani) [inline libs/api/domains/notifications/src/lib/notifications.resolver.ts:119]: This should be thrown by the client. If the client returns empty results but no error, can we really assume that there weren't any notifications? That seems presumptious

File: libs/api/domains/notifications/src/lib/notifications.resolver.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      throw e
+    }
+
+    if (!notification) {
+      this.logger.info('notification not found', {
+        notificationId,
+        locale,
+        category: LOG_CATEGORY,
+      })
+      throw new NotFoundException('notification not found')
```
Comment (thorkellmani) [inline libs/api/domains/notifications/src/lib/notifications.resolver.ts:83]: see below

File: libs/api/domains/notifications/src/lib/notifications.resolver.ts (hunk 2 above)
Comment (osk) [reply libs/api/domains/notifications/src/lib/notifications.resolver.ts:122]: Yes, it depends on how the actual client will be implemented, leaving as is for now

File: libs/api/domains/notifications/src/lib/notifications.resolver.ts (hunk 3 above)
Comment (osk) [reply libs/api/domains/notifications/src/lib/notifications.resolver.ts:85]: Yes, it depends on how the actual client will be implemented, leaving as is for now

**Comment by andes-it** [comment]
This PR's had a 15-day siesta. Lost its spark, did it? It's making an exit in 5 days unless the stale label disappears.

**Comment by andes-it** [comment]
This PR's had a 15-day siesta. Lost its spark, did it? It's making an exit in 5 days unless the stale label disappears.

File: libs/api/domains/notifications/src/lib/notifications.model.ts [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  ObjectType,
+  registerEnumType,
+} from '@nestjs/graphql'
+import { PageInfoDto } from '@island.is/nest/pagination'
+
+// TODO move to/import from client
+export enum NotificationStatus {
+  READ = 'READ',
+  UNREAD = 'UNREAD',
+}
```
Comment (thorkellmani) [inline libs/api/domains/notifications/src/lib/notifications.model.ts:15]: Todo?

File: libs/api/domains/notifications/src/lib/notifications.model.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+@ObjectType('Notification')
+export class Notification {
+  @Field(() => ID)
+  id!: number
+
+  @Field(() => ID)
+  notificationId!: string
+
```
Comment (thorkellmani) [inline libs/api/domains/notifications/src/lib/notifications.model.ts:77]: Graphql ID is supposed to be used for unique ids, like guids and stuff. Are these both that?

File: libs/api/domains/notifications/src/lib/notifications.model.ts [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+  })
+  after?: number
+
+  @Field({
+    nullable: true,
+    defaultValue: 10,
+  })
+  limit?: number
+}
+
```
Comment (thorkellmani) [inline libs/api/domains/notifications/src/lib/notifications.model.ts:116]: NestJS docs recommend specifying number types explicitly for number types. Either `() => Int`or `() => Float`

File: libs/api/domains/notifications/src/lib/notifications.module.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+@Module({
+  imports: [AuthModule, FeatureFlagModule, UserNotificationClientModule],
+  providers: [
+    NotificationsResolver,
+    NotificationsService,
+    {
+      provide: LOGGER_PROVIDER,
+      useValue: logger,
+    },
```
Comment (thorkellmani) [inline libs/api/domains/notifications/src/lib/notifications.module.ts:19]: This is probably unnecessary

File: libs/api/domains/notifications/src/lib/notifications.service.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    const notifications = await this.userNotificationsWAuth(
+      user,
+    ).meNotificationsControllerFindMany({
+      locale,
+      limit: input?.limit,
+      before: input?.first ? String(input?.first) : undefined,
+      after: input?.after ? String(input?.after) : undefined,
+    })
+
+    this.logger.info({ notifications })
```
Comment (thorkellmani) [inline libs/api/domains/notifications/src/lib/notifications.service.ts:45]: sus

File: libs/api/domains/notifications/src/lib/notifications.service.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    if (!notifications.data) {
+      this.logger.debug('no notification found')
+      return null
+    }
+
+    return {
+      data: notifications.data.map((item) => notificationMapper(item)),
+      messageCounts: {
+        totalCount: notifications.totalCount,
+        unreadCount: undefined, // TODO: Not currently returned
```
Comment (thorkellmani) [inline libs/api/domains/notifications/src/lib/notifications.service.ts:56]: todo?

**Review by thorkellmani — COMMENTED** [review]
Some thoughts

File: libs/api/domains/notifications/src/lib/notifications.service.ts (hunk 9 above)
Comment (thordurhhh) [reply libs/api/domains/notifications/src/lib/notifications.service.ts:56]: We're waiting on a service change for this. Keeping `undefined` until then.

File: libs/api/domains/notifications/src/lib/notifications.model.ts (hunk 5 above)
Comment (thordurhhh) [reply libs/api/domains/notifications/src/lib/notifications.model.ts:77]: Both are certainly unique. But `id` should just be a number in this case.

File: libs/api/domains/notifications/src/lib/notifications.model.ts (hunk 6 above)
Comment (thordurhhh) [reply libs/api/domains/notifications/src/lib/notifications.model.ts:116]: 🤝

File: libs/api/domains/notifications/src/lib/notifications.service.ts (hunk 8 above)
Comment (thordurhhh) [reply libs/api/domains/notifications/src/lib/notifications.service.ts:45]: yikes, my bad 🙈

File: libs/api/domains/notifications/src/lib/notifications.module.ts (hunk 7 above)
Comment (thorkellmani) [reply libs/api/domains/notifications/src/lib/notifications.module.ts:19]: why? you shouldn't need to provide the logger again 🦆

**Review by thorkellmani — APPROVED** [review]
cool lgtm 🏂

File: libs/api/domains/notifications/src/lib/notifications.model.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+  @Field(() => Int, {
+    nullable: true,
+    defaultValue: 10,
+  })
+  limit?: number
+}
+
+@ObjectType('Notifications')
+export class NotificationsResponse {
```
Comment (saevarma) [inline libs/api/domains/notifications/src/lib/notifications.model.ts:116]: There exists a helper function in `@island.is/nest/graphql` to create paginated response types

```suggestion
export class NotificationsResponse extends PaginatedResponse(Notification) { }
```

File: libs/api/domains/notifications/src/lib/notifications.model.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+  @Field(() => NotificationMessageCounts)
+  messageCounts!: NotificationMessageCounts
+
+  @Field(() => PageInfoDto)
+  pageInfo!: PageInfoDto
+}
+
+@ObjectType('NotificationResponse')
+export class NotificationResponse {
```
Comment (saevarma) [inline libs/api/domains/notifications/src/lib/notifications.model.ts:128]: Isn't the Response type wrapping unnecessary overhead? Same question for `MarkNotificationResponse`.

File: libs/api/domains/notifications/src/lib/notifications.model.ts [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+export class NotificationMessageCounts {
+  @Field(() => Int, { nullable: true })
+  totalCount?: number
+
+  @Field(() => Int, { nullable: true })
+  unreadCount?: number
+}
+
+@InputType('NotificationsInput')
+export class NotificationsInput {
```
Comment (saevarma) [inline libs/api/domains/notifications/src/lib/notifications.model.ts:99]: There exists a helper function in @island.is/nest/graphql to add the pagination input properties. And the [pagination chapter](https://docs.devland.is/technical-overview/api-design-guide/pagination#pagination-query-parameters) in API Design Guides describes those fields to be named, `limit`, `before`, `after`.

```suggestion
export class NotificationsInput extends PaginationInput() {
```

File: libs/api/domains/notifications/src/lib/notifications.module.ts (hunk 7 above)
Comment (saevarma) [reply libs/api/domains/notifications/src/lib/notifications.module.ts:19]: The `LoggingModule` provides the logger and exports it, so this isn't required.

File: libs/api/domains/notifications/src/lib/notifications.module.ts [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+import { logger, LOGGER_PROVIDER } from '@island.is/logging'
+import { UserNotificationClientModule } from '@island.is/clients/user-notification'
+
+import { Module } from '@nestjs/common'
+
+import { NotificationsResolver } from './notifications.resolver'
+import { NotificationsService } from './notifications.service'
+
+@Module({
+  imports: [AuthModule, FeatureFlagModule, UserNotificationClientModule],
```
Comment (saevarma) [inline libs/api/domains/notifications/src/lib/notifications.module.ts:12]: `AuthModule` isn't necessary here, as the root api app module loads it.
Looks like the `FeatureFlagModule` is unused?

File: libs/api/domains/notifications/src/lib/notifications.model.ts (hunk 10 above)
Comment (thordurhhh) [reply libs/api/domains/notifications/src/lib/notifications.model.ts:116]: Notification service uses offset based pagination. I was under the impression that the helper function was only for cursor based pagination.

Comment (thordurhhh) [reply libs/api/domains/notifications/src/lib/notifications.model.ts:108]: I am lying I was looking at the input.

File: libs/api/domains/notifications/src/lib/notifications.model.ts (hunk 12 above)
Comment (thordurhhh) [reply libs/api/domains/notifications/src/lib/notifications.model.ts]: Yes! good call. Updated.

**Review by kksteini — APPROVED** [review]
CODEOWNER files LGTM

File: libs/api/domains/notifications/src/lib/notifications.model.ts [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+export class NotificationsResponse extends PaginatedResponse(Notification) {
+  @Field(() => [Notification])
+  data!: Notification[]
+
+  @Field(() => Int, { nullable: true })
+  unreadCount?: number
+
+  @Field(() => PageInfoDto)
+  pageInfo!: PageInfoDto
+}
```
Comment (saevarma) [inline libs/api/domains/notifications/src/lib/notifications.model.ts:106]: The `PaginatedResponse` utility adds the `data`, `pageInfo` and `totalCount` fields to the model so you only needs to add the `unreadCount`
```suggestion
@ObjectType('Notifications')
export class NotificationsResponse extends PaginatedResponse(Notification) {
  @Field(() => Int, { nullable: true })
  unreadCount?: number
}
```

**Review by saevarma — APPROVED** [review]
🚀

File: libs/api/domains/notifications/src/lib/notifications.model.ts (hunk 14 above)
Comment (thordurhhh) [reply libs/api/domains/notifications/src/lib/notifications.model.ts]: 🤦

**Review by busla — APPROVED** [review]
codeowners LGTM

**Pull request merged**